### PR TITLE
Support a Solana address association feature

### PR DESCRIPTION
**Pull request by af-afk**

### CODE REVIEW

File: web/app.fluidity.money/app/routes/$network/dashboard/airdrop/index.tsx [hunk 1]
```diff
@@ -61,11 +62,25 @@ import Table, { IRow } from "~/components/Table";
...
+  PublicKey,
+  sendAndConfirmTransaction,
+} from "@solana/web3.js";
+import { Program, Idl, AnchorProvider, setProvider, web3 } from "@coral-xyz/anchor";
 
 const EPOCH_CURRENT_IDENTIFIER = "epoch_2";
 
 const AIRDROP_BLOG_POST = "https://blog.fluidity.money/announcing-the-fluidity-airdrop-and-ico-4c72172acb64";
 
+const AIRDROP_TGE_CLAIM = "http://app.fluidity.money/arbitrum/dashboard/airdrop#recap"
```
Comment (eli-d) [inline web/app.fluidity.money/app/routes/$network/dashboard/airdrop/index.tsx]: use a relative redirect

File: web/app.fluidity.money/app/routes/$network/dashboard/airdrop/index.tsx [hunk 2]
```diff
@@ -289,20 +305,131 @@ const Airdrop = () => {
     };
 
     const YouAreEligible = () => {
+      // used to track if the button was clicked to delegate. not disabled.
+      // the user will have to reload the page to set it again. unless an error happened!
+      const [beginAssociation, setBeginAssociation] = useState(false);
```
Comment (eli-d) [inline web/app.fluidity.money/app/routes/$network/dashboard/airdrop/index.tsx:301]: could encapsulate association status into a single state like `"pending" | "inProgress" | "completed"`

File: web/app.fluidity.money/app/routes/$network/dashboard/airdrop/index.tsx [hunk 3]
```diff
@@ -289,20 +305,131 @@ const Airdrop = () => {
...
+      // used to paint the button backdrop to set the disabled
+      const [validAssociationAddress, setValidAssociationAddress] = useState(false);
+
+      // used to prevent double submitting until the request is done
+      const [airdropAssociationInProgress, setAirdropAssociationInProgress] = useState(false);
+
+      // the ethereum address they want to associate
+      const [associationEthereumAddress, setAssociationEthereumAddress] = useState("");
+
+      const [hasEthereumAddressChangedSinceClick, setHasEthereumAddressChangedSinceClick] =
```
Comment (eli-d) [inline web/app.fluidity.money/app/routes/$network/dashboard/airdrop/index.tsx]: this doesn't seem to ever be used?

File: web/app.fluidity.money/app/routes/$network/dashboard/airdrop/index.tsx [hunk 4]
```diff
@@ -289,20 +305,131 @@ const Airdrop = () => {
...
+        // process of using anchor to send out the transaction using
+        // Phantom that associates the address.
+
+        const s = associationEthereumAddress;
+
+        setHasEthereumAddressChangedSinceClick(false);
+
+        // regex taken from the transfer route
+        const addressRegex = /(0x[a-fA-F0-9]{40})/g;
+        const validAddr = s.match(addressRegex) || ethers.utils.isAddress(s);
```
Comment (eli-d) [inline web/app.fluidity.money/app/routes/$network/dashboard/airdrop/index.tsx:351]: why do both of these?

**Review by eli-d — APPROVED** [review]

File: web/app.fluidity.money/app/routes/$network/dashboard/airdrop/index.tsx (hunk 4 above)
Comment (af-afk) [reply web/app.fluidity.money/app/routes/$network/dashboard/airdrop/index.tsx:351]: I took this from the existing send/receive implementation. I'm just going to assume it's fine for now.

**Pull request merged**